src/pybind/mgr/volumes/fs_subvolume.py

```diff
@@ -0,0 +1,365 @@
+"""
+Copyright (C) 2019 Red Hat, Inc.
+
+LGPL2.1.  See file COPYING.
+"""
+
+import errno
+import logging
+import os
+
+import cephfs
+import rados
+
+def to_bytes(param):
+    '''
+    Helper method that returns byte representation of the given parameter.
+    '''
+    if isinstance(param, str):
+        return param.encode()
+    else:
+        return str(param).encode()
+
+log = logging.getLogger(__name__)
+
+# Reserved subvolume group name which we use in paths for subvolumes
+# that are not assigned to a group (i.e. created with group=None)
+NO_GROUP_NAME = "_nogroup"
+
+
+class SubvolumePath(object):
+    """
+    Identify a subvolume's path as group->subvolume
+    The Subvolume ID is a unique identifier, but this is a much more
+    helpful thing to pass around.
+    """
+    def __init__(self, group_id, subvolume_id):
+        self.group_id = group_id
+        self.subvolume_id = subvolume_id
+        assert self.group_id != NO_GROUP_NAME
+        assert self.subvolume_id != "" and self.subvolume_id is not None
+
+    def __str__(self):
+        return "{0}/{1}".format(self.group_id, self.subvolume_id)
+
+
+class SubvolumeClient(object):
+    """
+    Combine libcephfs and librados interfaces to implement a
+    'Subvolume' concept implemented as a cephfs directory.
+
+    Additionally, subvolumes may be in a 'Group'.  Conveniently,
+    subvolumes are a lot like manila shares, and groups are a lot
+    like manila consistency groups.
+
+    Refer to subvolumes with SubvolumePath, which specifies the
+    subvolume and group IDs (both strings).  The group ID may
+    be None.
+
+    In general, functions in this class are allowed raise rados.Error
+    or cephfs.Error exceptions in unexpected situations.
+    """
+
+    # Where shall we create our subvolumes?
+    DEFAULT_SUBVOL_PREFIX = "/subvolumes"
+    DEFAULT_NS_PREFIX = "fssubvolumens_"
```

Comment:
Did these names change from `ceph_volume_client.py`? How are we going to import existing manila volumes?

Reply:
Good catch. I'll set this back to 

> > In Manila's CephFS driver and CSI's CephFS driver, currently the removal of subvolume is idempotent in that if subvolume is already removed, ENOENT isn't raised.
> > See,
> > https://github.com/ceph/ceph-csi/blob/csi-v1.0/pkg/cephfs/volume.go#L111
> > I think I'll revert to implementing this behavior, and not raise 'ENOENT'.
> 
> The driver can treat ENOENT as success and not bubble the error up. Wouldn't that suffice?

That might work too

Reply:
Yes. Good catch!

I've reverted to the defaults to the ones in ceph_volume_client.py for now,
DEFAULT _SUBVOL_PREFIX="/volumes"
DEFAULT_NS_PREFIX="fsvolume_ns"

So the the default subvolume path would look like as is the case for existing manila volumes,
/volumes/_nogroup/<subvol_name>

Is that OK? Sage had suggested that it be, /subvolumes/<foo>/<bar>

The other option would be to be to allow have the new defaults,
DEFAULT _SUBVOL_PREFIX="/volumes"
DEFAULT_NS_PREFIX="fsvolume_ns"

And allow changing the subvol prefix and ns prefix through `fs subvolume` command like,
```
ceph fs subvolume create volume00 subvolume00 --subvol_prefix  volumes  --namespace_prefix fsvolumens_ 
```
In the module.py we could then pass these,
 ```
def _fs_subvolume_create(self, inbuf, cmd):
      ...
      with SubvolumeClient(self, subvolume_prefix=subvol_prefix, pool_ns_prefix=namespace_prefix, fs_name=vol_name, ) as svc:

```

Reply:
I'm for allowing the admin/operator to change the `subvol_prefix` but I would think that'd be a `ceph fs volume` command as it would apply to **all** subvolumes?

I don't like the idea of users making subvolumes anywhere they want. At some point, we will want to protect the `ceph fs volume` commands so that only the operator can change things. Then users can be given permission to create subvolumes via certain cephx authorizations.

Reply:
As for the name itself, I'm okay with changing it but there should be a way to automagically import the older volumes. Ideally that should just be a series of `rename` syscalls.